### PR TITLE
Txn persistence pt1

### PR DESCRIPTION
**Pull request by iafuture**

### CODE REVIEW

File: src/k2/module/k23si/Module.cpp [hunk 1]
```diff
@@ -835,90 +918,114 @@ bool K23SIPartitionModule::_makeProjection(dto::SKVRecord::Storage& fullRec, dto
     return true;
 }
 
+template<typename ResponseT>
+seastar::future<std::tuple<Status, ResponseT>>
+K23SIPartitionModule::_respondAfterFlush(Status&& status, ResponseT&& response) {
+    return _persistence->flush()
+        .then([status=std::move(status), response=std::forward<ResponseT>(response)] (auto&& flushStatus) mutable {
+            if (flushStatus.is2xxOK())
```
Comment (iafuture) [inline src/k2/module/k23si/Module.cpp:926]: fip to negative check

File: src/k2/module/k23si/Module.cpp [hunk 2]
```diff
@@ -835,90 +918,114 @@ bool K23SIPartitionModule::_makeProjection(dto::SKVRecord::Storage& fullRec, dto
...
-        // do not allow empty partition key
-        return RPCResponse(dto::K23SIStatus::BadParameter("missing partition key in write"), dto::K23SIWriteResponse{});
+    // NB: failures in processing a write do not require that we set the TR state to aborted at the TRH. We rely on
+    //     the client to do the correct thing and issue an abort on a failure.
+    K2LOG_D(log::skvsvr, "Partition: {}, handle write: {}", _partition, request);
+    if (request.designateTRH) {
+        K2LOG_D(log::skvsvr, "Partition: {}, designating TRH in {}", _partition, request.mtr);
+        return _designateTRH({.trh=request.trh, .mtr=request.mtr})
+            .then([this, request=std::move(request), deadline] (auto&& status) mutable {
+                if (status.is2xxOK()) {
```
Comment (iafuture) [inline src/k2/module/k23si/Module.cpp:971]: flip to negative check

File: src/k2/module/k23si/Persistence.h [hunk 3]
```diff
@@ -32,24 +32,46 @@ namespace k2 {
...
+        append(val);
+
+        _pendingProms.emplace_back();
+        return _pendingProms.back().get_future();
+    }
+
+    // Append the given value to the persistence buffer. An explicit call to flush() is needed to send the data out
+    template<typename ValueType>
+    void append(const ValueType& val) {
+        if (!_buffer) {
```
Comment (iafuture) [inline src/k2/module/k23si/Persistence.h:65]: check for stopped

File: src/k2/module/k23si/Persistence.cpp [hunk 4]
```diff
@@ -32,4 +32,50 @@ Persistence::Persistence() {
...
+seastar::future<> Persistence::stop() {
+    _stopped = true;
+    K2LOG_D(log::skvsvr, "Stopping with {} pending promises", _pendingProms.size());
+
+    return flush().discard_result();
+}
+
+seastar::future<Status> Persistence::flush() {
+    K2LOG_D(log::skvsvr, "flush with bs={} and {} pending promises", (_buffer? _buffer->getSize() : 0), _pendingProms.size());
+    if (!_buffer) {
```
Comment (iafuture) [inline src/k2/module/k23si/Persistence.cpp:44]: check for pending promises

File: src/k2/module/k23si/Module.cpp [hunk 5]
```diff
@@ -835,90 +918,117 @@ bool K23SIPartitionModule::_makeProjection(dto::SKVRecord::Storage& fullRec, dto
...
-    auto& versions = _indexer[request.key];
-    // in this situation, return AbortRequestTooOld error.
-    {
-        Status validateStatus = _validateStaleWrite(request, versions);
-        if (!validateStatus.is2xxOK()) {
-            K2LOG_D(log::skvsvr, "Partition: {}, request too old for key {} due to {}",
-                    _partition, request.key, validateStatus);
-            return RPCResponse(std::move(validateStatus), dto::K23SIWriteResponse{});
+                K2LOG_D(log::skvsvr, "Partition: {}, succeeded creating TR. Processing write for {}", _partition, request.mtr);
+                request.designateTRH = false;
```
Comment (jerryhfeng) [inline src/k2/module/k23si/Module.cpp:978]: Why do we unset the flag? Is it needed at all?

File: src/k2/module/k23si/Module.cpp [hunk 6]
```diff
@@ -835,90 +918,117 @@ bool K23SIPartitionModule::_makeProjection(dto::SKVRecord::Storage& fullRec, dto
...
+                    // challenger must fail. Flush in case a TR was created during this call to handle write
+                    K2LOG_D(log::skvsvr, "Partition: {}, write push challenger lost for key {}", _partition, request.key);
+                    return _respondAfterFlush(dto::K23SIStatus::AbortConflict("incumbent txn won in write push"), dto::K23SIWriteResponse{});
                 }
-                // challenger must fail
-                K2LOG_D(log::skvsvr, "Partition: {}, write push challenger lost for key {}", _partition, request.key);
-                return RPCResponse(dto::K23SIStatus::AbortConflict("incumbent txn won in write push"), dto::K23SIWriteResponse{});
+
+                K2LOG_D(log::skvsvr, "Partition: {}, write push retry for key {}", _partition, request.key);
+                return handleWrite(std::move(request), deadline);
```
Comment (jerryhfeng) [inline src/k2/module/k23si/Module.cpp:1015]: Maybe should call directly _processWrite()?

**Review by jerryhfeng — DISMISSED** [review]
didn't check TxnManager.cc code as they(transition map) will be changed again soon.

A couple minor questions.

Otherwise, looks good to me.

File: src/k2/module/k23si/Module.cpp (hunk 5 above)
Comment (iafuture) [reply src/k2/module/k23si/Module.cpp:978]: No, it isn't needed. It was there before so I kept it in, but we can remove

File: src/k2/module/k23si/Module.cpp (hunk 6 above)
Comment (iafuture) [reply src/k2/module/k23si/Module.cpp:1015]: yup

**Review by jfunston — APPROVED** [review]

**Pull request merged**